site/en/docs/extensions/mv3/tut_analytics/index.md

```diff
@@ -1,166 +1,279 @@
 
-```js
-(function() {
-  var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
-  ga.src = 'https://ssl.google-analytics.com/ga.js';
-  var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
-})();
-```
+### Generate a `client_id` {: #generate-client-id }
 
-You'll also need to ensure that your extension has access to load the resource by relaxing the
-default content security policy. The policy definition in your [`manifest.json`][6] might look like:
+The second step is to generate a unique identifier for a specific device/user, the `client_id`. The id should stay the same, as long as the extension is installed on a user’s browser. It can be an arbitrary string, but should be unique to the client. You can generate one by calling [`self.crypto.randomUUID()`][7]. Store the `client_id` in [`chrome.storage.local`][8] to make sure it stays the same as long as the extension is installed.
+
+Using `chrome.storage.local` requires the `storage` permission in your manifest file:
+
+{% Label %}manifest.json:{% endLabel %}
 
 ```json
 {
-  ...,
-  "content_security_policy": {
-    "extension_pages": "script-src 'self' https://ssl.google-analytics.com; object-src 'self'",
-  },
-  ...
+  …
+  "permissions": ["storage"],
+  …
 }
 ```
 
-Here is a popup page (`popup.html`) which loads the asynchronous tracking code via an external
-JavaScript file (`popup.js`) and tracks a single page view:
+Then you can use `chrome.storage.local` to store the `client_id`:
 
-```html
-<!DOCTYPE html>
-<html>
- <head>
-   ...
-  <script src="popup.js"></script>
- </head>
- <body>
-   ...
- </body>
-</html>
+```js
+async function getOrCreateClientId() {
+  const result = await chrome.storage.local.get('clientId');
+  let clientId = result.clientId;
+  if (!clientId) {
+    // Generate a unique client ID, the actual value is not relevant
+    clientId = self.crypto.randomUUID();
+    await chrome.storage.local.set({clientId});
+  }
+  return clientId;
+}
 ```
 
+### Send an analytics event {: #send-analytics-events }
+
+With the API credentials and the `client_id`, you can send an event to Google Analytics via a [`fetch`][9] request:
+
 ```js
-var _gaq = _gaq || [];
-_gaq.push(['_setAccount', 'UA-XXXXXXXX-X']);
-_gaq.push(['_trackPageview']);
-
-(function() {
-  var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
-  ga.src = 'https://ssl.google-analytics.com/ga.js';
-  var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
-})();
+const GA_ENDPOINT = 'https://www.google-analytics.com/mp/collect';
+const MEASUREMENT_ID = `G-...`;
+const API_SECRET = `...`;
+
+fetch(
+  `${GA_ENDPOINT}?measurement_id=${MEASUREMENT_ID}&api_secret=${API_SECRET}`,
+  {
+    method: 'POST',
+    body: JSON.stringify({
+      client_id: getOrCreateClientId(),
```

Comment:
Is this missing an await?

Reply:
oooh yes - all of them are, added - thanks!